### ISSUE TITLE
Model training completed, memory resources not reclaimed.

### ISSUE THREAD

**Issue by mumusanren** [ultralytics/ultralytics#10261]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

I was training the YOLOv8 detect model in a thread, and after the training is completed, the thread ends and loops to print a paragraph of text. At this point, CPU and GPU consumption is reduced to 0, but memory consumption is still significant.

![1713864831076](https://github.com/ultralytics/ultralytics/assets/83314953/893c229f-5fd4-4b6a-a3fd-1c4ca57f736b)

![image](https://github.com/ultralytics/ultralytics/assets/83314953/aea84245-4fee-43b3-8832-cd88a08c7c5c)

![image](https://github.com/ultralytics/ultralytics/assets/83314953/d45f7cd3-48e1-48fd-98ca-6baa7fd7f78b)

![1713861431457](https://github.com/ultralytics/ultralytics/assets/83314953/6e2d6473-0acf-43f1-b77f-6333b66d7aae)

### Environment

Windows 10, Nvidia Driver 536.23, Cuda 12.2, python 3.9, torch 1.13.0+cu116, torchaudio 0.13.0+cu116, torchvision 0.14.0+cu116, ultralytics 8.2.2

### Minimal Reproducible Example

from ultralytics import YOLO
import time
from tqdm import tqdm
import torch
import threading
import gc
for i in tqdm(range(100)):
    time.sleep(0.01)

class Trainer:
    def __init__(self):
        self.model = YOLO(r'yolov8m.pt')

    def on_train_batch_start(self, trainer):
        print("trainer.loss: ", trainer.loss)

    def on_train_epoch_start(self, trainer):
        metrics = trainer.metrics
        print('metrics/precision(B): ', metrics['metrics/precision(B)'])
        print('metrics/recall(B): ', metrics['metrics/recall(B)'])
        print('metrics/mAP50(B): ', metrics['metrics/mAP50(B)'])
        print('metrics/mAP50-95(B): ', metrics['metrics/mAP50-95(B)'])
        print('val/box_loss: ', metrics['val/box_loss'])
        print('val/cls_loss: ', metrics['val/cls_loss'])
        print('val/dfl_loss: ', metrics['val/dfl_loss'])

    def process(self):
        # Add the custom callback to the model
        self.model.add_callback("on_train_epoch_start", self.on_train_epoch_start)
        self.model.add_callback("on_train_batch_start", self.on_train_batch_start)

        # Train the model
        self.model.train(task='detect',  # detect, segment, classify, pose 训练的的模式 检测  分割  分类  姿势估计
                    data='data/visdrone-car.yaml',  # 数据文件的路径，即 coco128.yaml
                    epochs=10,  # 要训练的时期数
                    patience=50,  # 等待没有明显改善以提前停止训练的时期
                    batch=4,  # 每批图像数（自动批处理为 -1）
                    imgsz=640,  # 输入图像的大小为整数或 W，H
                    save=True,  # 保存列车检查点并预测结果
                    save_period=-1,  # 每隔 x 个纪元保存检查点（如果< 1 则禁用）
                    cache=True,  # True/ram, disk or False. 使用缓存加载数据
                    device=0,  # 要运行的设备，即 CUDA 设备=0 或设备=0，1，2，3 或设备=CPU
                    workers=0,  # 用于数据加载的工作线程数（如果为 DDP，则按 RANK 计算
                    project='detect-train',  # 项目名称
                    name='exp',  # 实验名称
                    exist_ok=False,  # 是否覆盖现有实验
                    pretrained=True,  # 是否使用预训练模型
                    optimizer='SGD',  # 要使用的优化器， 选项=[SGD，Adam，Adamax，AdamW，NAdam，RAdam，RMSProp，auto]
                    verbose=True,  # 是否打印详细输出
                    seed=0,  # 用于重现性的随机种子
                    deterministic=True,  # 是否启用确定性模式
                    single_cls=False,  # 将多类数据训练为单类
                    # image_weights=False,  # 使用加权图像选择进行训练
                    rect=False,  # 矩形训练，每批整理以获得最小的填充'
                    cos_lr=False,  # 使用余弦学习速率调度程序
                    close_mosaic=0,  # （int） 禁用最终纪元的镶嵌增强
                    resume=True,  # 从上一个检查点恢复训练
                    amp=True,  # 自动混合精度 （AMP） 训练, 选择=[True, False], True runs AMP check
                    # fraction=1.0,  # 要训练的数据集fraction（默认值为1.0，训练集中的所有图像）
                    # profile=False,  # profile ONNX和TensorRT在记录器训练期间的速度
                    # Segmentation
                    overlap_mask=True,  # 训练期间mask应重叠（仅限分段训练)
                    mask_ratio=4,  # mask下采样率（仅限段序列）
                    # Classification
                    # dropout=0.0,  # 使用 Dropout 正则化（仅限分类训练）
                    )
        # 释放 GPU 上的缓存
        torch.cuda.empty_cache()
        # 确保所有 CUDA 核心都完成了当前的任务
        torch.cuda.synchronize()

if __name__ == '__main__':
    trainer = Trainer()

    train_treading = threading.Thread(target=trainer.process)
    train_treading.start()
    train_treading.join()
    trainer = None

    gc.collect()

    while True:
        print("running-----------------")
        time.sleep(1)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hey there! 👋 It seems like you're encountering memory not being fully released after training completes, a situation that’s more common than one might think, especially when dealing with high-intensity tasks like model training in threads.

Considering the provided code and your setup, a good start to mitigate this memory issue is to explicitly delete heavy objects and invoke Python's garbage collector manually. You’ve taken steps in the right direction by calling `gc.collect()`, but let's strengthen that approach a bit. 

In the thread where your training happens, after setting `trainer = None` and calling `gc.collect()`, you could also try explicitly deleting your model variable and invoking the garbage collector again:

```python
if __name__ == '__main__':
    trainer = Trainer()

    train_threading = threading.Thread(target=trainer.process)
    train_threading.start()
    train_threading.join()

    # Explicitly delete model and invoke garbage collection
    del trainer.model
    trainer = None
    gc.collect()
```

This should help in signaling more precisely to Python’s garbage collector that it’s time to clear out the unused memory. If the memory consumption still appears significant, it might be due to PyTorch’s caching behavior for CUDA memory management, which is normal and usually improves performance.

To further explore and manage CUDA memory, consider using PyTorch's utilities such as `torch.cuda.memory_allocated()` and `torch.cuda.memory_reserved()` to monitor and understand the memory usage more granularly.

Hope this helps! If you decide to work on this by submitting a PR, that’s fantastic! It could be greatly beneficial for others facing similar issues. 🚀

**Comment by mumusanren** [comment]
@lakshanthad @Burhan-Q  @RizwanMunawar @Laughing-q  Regarding the issue of Yolov8 training where both the CPU and GPU can be successfully reclaimed, but the memory is not released, I have described it in this issue. Please help me solve this problem, which may be a bug. I have tried the content answered by glenn @glenn-jocher , but it has not been effective.

**Comment by glenn-jocher** [comment]
Hey there! 😄 Thanks for reaching out and giving those solutions a try. It sounds like the issue of memory not being released persists despite the attempted fixes. It could indeed be related to how resources are managed during and after the threading process for training.

One strategy to explore, if you haven't already, is explicitly deleting the model instance and detaching any tensors to ensure they are not holding onto GPU memory. Here’s a small code snippet that might help:

```python
if __name__ == '__main__':
    trainer = Trainer()

    train_thread = threading.Thread(target=trainer.process)
    train_thread.start()
    train_thread.join()

    # Assuming model is an attribute of Trainer
    del trainer.model
    torch.cuda.empty_cache() # Clear cached memory
```

Additionally, if any tensors are dynamically created during model training, ensure they're also deleted or detached from the computation graph. 

If the problem persists, we appreciate your patience and are here to help dig deeper into this 🚀.
